src/nameck.rs

```diff
@@ -329,10 +329,31 @@ impl Nameset {
             .expect("please only use get_atom for local $v")
     }
 
-    /// Returns the last known atom
+    /// Returns an atom uniquely identifiable by the given `u32`, 
+    /// and guaranteed to be different of any "regular" atom
+    /// previously allocated during parsing.
+    #[inline]
+    #[must_use]
+    pub fn create_extra_atom(&self, id: u32) -> Option<Atom> {
+        let intern = u32::MAX.saturating_sub(id);
+        if intern > self.atom_table.table.len() as u32 { Some(Atom(intern)) }
+        else { None }
+    }
+
+    /// Retrieves the unique `u32` identifier given when creating an "extra" atom,
+    /// typically for work variables
+    #[inline]
+    #[must_use]
+    pub fn extra_atom_id(&self, atom: Atom) -> Option<u32> {
+        if self.is_extra_atom(atom) { Some(u32::MAX.saturating_sub(atom.0)) }
+        else { None }
+    }
+
+    /// Tests whether an atom has been allocated using `create_extra_atom`.
+    #[inline]
     #[must_use]
-    pub fn last_atom(&self) -> Atom {
-        Atom(self.atom_table.table.len() as u32)
+    pub fn is_extra_atom(&self, atom: Atom) -> bool {
```

Comment:
I'm not sure that we should promise that extra atoms always come in a block after regular atoms, or even that there is a way to distinguish them. What if the parser is restarted after some extra atoms are created, in incremental use cases?

Reply:
So right now I'm kind of eating the cake starting from both sides:
`Nameset` naturally allocates atoms from 0 up, continuously, while I'm getting the identifiers from MAX down, assuming the passed ID stays small.

What my client currently does to compute the ID, is that it also ranks them and assumes that small indices are used first (`&W1`, `&S1`, `&C1`, `&W2`, `&S2`, `&C2`, `&W3`, etc.) This allows a direct computation without any lookup.

There are currently around 100'000 atoms (99'577) in set.mm, so there is still plenty of space until 2^32.

Beyond the fact that this method works with direct computations and no lookup, which shall be faster, it most importantly works with an immutable `Nameset` instance. If `Nameset` were to reserve internal spots for the "extra" atoms, this means it needs to be mutated, which goes against the current philosophy of "once the DB is parsed, everything is immutable", which allows very fast execution since we only pass around `Arc`s to the different structures of the DB.

Let's assume we went up to work variable `&W99999`, and that each time the parser is restarted everything changes and it requires a new set of 100'000 atom indices. Using `u32` as currently, the parsing could be restarted 43'000 times until there is a clash.

Do you have another idea?

Reply:
> Beyond the fact that this method works with direct computations and no lookup, which shall be faster, it most importantly works with an immutable `Nameset` instance. If `Nameset` were to reserve internal spots for the "extra" atoms, this means it needs to be mutated, which goes against the current philosophy of "once the DB is parsed, everything is immutable", which allows very fast execution since we only pass around `Arc`s to the different structures of the DB.

I think that a name generator that does not mutate can't possibly work correctly. I think it is fine to use interior mutability here, and the only state a name generator needs to track is the last allocated ID so a `AtomicU32` will suffice, so API-wise it can still take a `&self`.

I'm still a little dubious about this mmj2-styled work variable thing. This really ought to be some kind of enum, work vars are not the same as regular vars in a whole bunch of different ways, and adding phantom `$f` statements to the database seems like a really weird encoding to me.